File: 백준/Silver/1389. 케빈 베이컨의 6단계 법칙/케빈 베이컨의 6단계 법칙.cc
#include <bits/stdc++.h>
using namespace std;
int n, m, a, b, ans, cnt, dis[101];
vector<int> adj[101];
int main(){
    cin >> n >> m;
    for (int i = 0; i < m; i++) {
        cin >> a >> b;
        adj[a].push_back(b);
        adj[b].push_back(a);
    }
    ans = n + 1, cnt = (n + 1) * (n + 1);
    for (int i = 1; i < n + 1; i++) {
        deque< pair<int, int> > dq;
        for(int j = 0; j < n + 1; j++)
            dis[j] = n + 1;
        dq.push_back(make_pair(i, 0));
        while (!dq.empty()){
            pair<int, int> cur = dq.front();
            dq.pop_front();
            for(int pos : adj[cur.first]){
                if (dis[pos] > cur.second + 1){
                    dq.push_back(make_pair(pos, cur.second + 1));
                    dis[pos] = cur.second + 1;
                }
            }
        }
        if (accumulate(dis + 1, dis + n + 1, 0) < cnt){
            cnt = accumulate(dis + 1, dis + n + 1, 0);
            ans = i;
        }
    }
    cout << ans;
    return 0;
}

File: 백준/Silver/1389. 케빈 베이컨의 6단계 법칙/케빈 베이컨의 6단계 법칙.py
import sys
from collections import deque

n, m = map(int, sys.stdin.readline().rstrip().split())
adj = [[] for _ in range(n + 1)]
for i in range(m):
    a, b = map(int, sys.stdin.readline().rstrip().split())
    adj[a].append(b)
    adj[b].append(a)
ans, cnt = n + 1, (n + 1) * (n + 1)
for i in range(1, n + 1):
    distance, dq = [n + 1 for _ in range(n + 1)], deque()
    distance[i] = 0
    dq.appendleft((i, 0))
    while dq:
        cur, dis = dq.pop()
        for a in adj[cur]:
            if distance[a] > dis + 1:
                distance[a] = dis + 1
                dq.appendleft((a, dis + 1))
    if sum(distance[1:]) < cnt:
        cnt = sum(distance[1:])
        ans = i

print(ans)
